Realm/Tests/RealmTests.mm

```diff
@@ -1455,6 +1455,381 @@ - (void)testActiveVersionLimit {
+    __block bool notificationFired = false;
+    RLMNotificationToken *token = [realm addNotificationBlock:^(__unused NSString *note, RLMRealm *realm) {
+        XCTAssertEqual(1U, [StringObject allObjectsInRealm:realm].count);
+        notificationFired = true;
+    }];
+
+    [realm asyncTransactionWithBlock:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+    } onComplete:^{
+        [asyncComplete fulfill];
+    }];
+
+    [self waitForExpectationsWithTimeout:3.0 handler:nil];
+    // notification shouldnt have fired
+    XCTAssertTrue(notificationFired);
+    [token invalidate];
+}
+
+- (void)testBeginAsyncTransactionInAsyncTransaction {
+    RLMRealm *realm = [self realmWithTestPath];
+    XCTestExpectation *transaction1 = [self expectationWithDescription:@"async transaction 1 complete"];
+    XCTestExpectation *transaction2 = [self expectationWithDescription:@"async transaction 2 complete"];
+
+    [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        
+        [realm beginAsyncWriteTransaction:^{
+            [StringObject createInRealm:realm withValue:@[@"string"]];
+            
+            [realm commitAsyncWriteTransaction:^{
+                XCTAssertEqual(0, [StringObject allObjects].count);
+                [transaction1 fulfill];
+            }];
+        }];
+        [realm commitAsyncWriteTransaction:^{
+            XCTAssertEqual(0, [StringObject allObjects].count);
+            [transaction2 fulfill];
+        }];
+    }];
+    
+    [self waitForExpectationsWithTimeout:3.0 handler:nil];
+    XCTAssertEqual(0, [StringObject allObjects].count);
+}
+
+- (void)testAsyncTransactionFromSyncTransaction {
+    RLMRealm *realm = [self realmWithTestPath];
+    XCTestExpectation *transaction1 = [self expectationWithDescription:@"async transaction 1 complete"];
+
+    [realm beginWriteTransaction];
+    [StringObject createInRealm:realm withValue:@[@"string"]];
+
+    [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+
+        [realm commitAsyncWriteTransaction:^{
+            XCTAssertEqual(0, [StringObject allObjects].count);
+            [transaction1 fulfill];
+        }];
+    }];
+    [realm commitWriteTransaction];
+    
+    [self waitForExpectationsWithTimeout:3.0 handler:nil];
+    XCTAssertEqual(0, [StringObject allObjects].count);
+}
+
+- (void)testAsyncTransactionCancel {
+    RLMRealm *realm = [RLMRealm defaultRealm];
+    auto expectation = [self expectationWithDescription:@""];
+    expectation.expectedFulfillmentCount = 2;
+
+    [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        [expectation fulfill];
+    }];
+
+    AsyncHandle handle = [realm beginAsyncWriteTransaction:^{
+        [StringObject createInRealm:realm withValue:@[@"string"]];
+        [expectation fulfill];
+        [realm cancelAsyncTransaction:handle];
```

Comment:
Does this actually compile?

Reply:
Yes, and working perfectly. It was an issue with variable capturing by the closure in the Swift implementation but it's OK for objc.